internal/pkg/client/cache/dir.go

```diff
@@ -22,42 +23,186 @@ const (
 	// for image downloads to be cached in
 	DirEnv = "SINGULARITY_CACHEDIR"
 
-	// cacheDir specifies the name of the directory relative to the
+	// CacheDir specifies the name of the directory relative to the
 	// singularity data directory where images are cached in by
 	// default.
 	// Uses "~/.singularity/cache" which will not clash with any 2.x cache
 	// directory.
-	cacheDir = "cache"
+	CacheDir = "cache"
 )
 
-var root string
+// ImgCache is an structure representing a cache
+type ImgCache struct {
+	// Basedir is the base directory of the image cache. By default, it is set
+	// to $HOME/.singularity. Users can also set the SINGULARITY_CACHEDIR
+	// environment variable to set BaseDir. BaseDir is also used when the code
+	// sets the SINGULARITY_CACHEDIR for a child process that needs to use the
+	// image cache.
+	BaseDir string
```

Comment:
I _still_ think this should not be exported.

I was going to comment this elsewhere, but it probably makes more sense to do it here.

The problem with exporting all these fields instead of having functions to access the values is this:

```
package cache

type ImgCache struct {
    Library string
}
```

and then later:

```
package foo

func bar() {
    c := ImgCache{Library: "some value"}
}

func quux() {
    h := getCacheHandle()
    h.Library = "some value"
}
```

if you do this instead:

```
package cache

type ImgCache struct {
    library string
}
```

and then later:

```
package foo

func bar() {
    c := ImgCache{library: "some value"} // this is a compilation error
}

func quux() {
    h := getCacheHandle()
    h.library = "some value" // this is a compilation error
}
```

also, since all the fields in the struct would not be exported, and since the zero value of the struct is not useful, users are forced to go thru `cache.NewHandle` to obtain a handle. Of course within the cache package you can still do all sort of nasty stuff, because it would have access to all the unexported fields.

The following does not need to be implemented, I'm mentioning it for the rationale.

Defining functions like this also allows you to define interfaces. For example:

```
package shub

type ShubCache interface {
    Shub() string
}

func Download(cache ShubCache, ...) {
    // ...
    dir := cache.Shub()
    // ...
}
```

and other code could pass an instance of cache.Handle to this function.

Reply:
Not possible: we need `BaseDir` to correctly set the environment variable when starting a new process via `exec` as the comments are stating. We can have a `GetBasedir()` function tho. @mem would you prefer that option? Your point about not letting anyone change the value of `BaseDir` is very valid.

Reply:
I'm sorry I'm not being clear: change `BaseDir` to `baseDir` and add a function to _read_ the value. I'm not asking you to remove it.